File: scripts/download_tiles.py
import math
import os
import requests
import time

def deg2num(lat_deg, lon_deg, zoom):
  lat_rad = math.radians(lat_deg)
  n = 2.0 ** zoom
  xtile = int((lon_deg + 180.0) / 360.0 * n)
  ytile = int((1.0 - math.asinh(math.tan(lat_rad)) / math.pi) / 2.0 * n)
  return (xtile, ytile)

# Bounding box covering Leon, Iloilo and the paths
min_lat = 10.74
max_lat = 10.91
min_lon = 122.28
max_lon = 122.43

zoom_levels = [11, 12, 13, 14]
base_dir = os.path.join(os.path.dirname(__file__), "static", "tiles")

headers = {
    'User-Agent': 'OfflineEvacMap/1.0 (Contact: student@example.com)'
}

def download_tiles():
    total_downloaded = 0
    for z in zoom_levels:
        x_min, y_max = deg2num(min_lat, min_lon, z)
        x_max, y_min = deg2num(max_lat, max_lon, z)
        
        if x_min > x_max: x_min, x_max = x_max, x_min
        if y_min > y_max: y_min, y_max = y_max, y_min
        
        print(f"Zoom {z}: X {x_min}-{x_max}, Y {y_min}-{y_max}")
        for x in range(x_min, x_max + 1):
            for y in range(y_min, y_max + 1):
                tile_dir = os.path.join(base_dir, str(z), str(x))
                os.makedirs(tile_dir, exist_ok=True)
                tile_path = os.path.join(tile_dir, f"{y}.png")
                
                if not os.path.exists(tile_path):
                    url = f"https://tile.openstreetmap.org/{z}/{x}/{y}.png"
                    try:
                        resp = requests.get(url, headers=headers)
                        if resp.status_code == 200:
                            with open(tile_path, 'wb') as f:
                                f.write(resp.content)
                            total_downloaded += 1
                        else:
                            print(f"Failed to download {url}: {resp.status_code}")
                    except Exception as e:
                        print(f"Error downloading {url}: {e}")
                    time.sleep(0.1) # Be polite to OSM
                    
    print(f"Tile download complete. Downloaded {total_downloaded} new tiles.")

if __name__ == '__main__':
    download_tiles()

File: www/download_tiles.js
const fs = require('fs');
const https = require('https');
const path = require('path');

const z = 17;
const minLat = 10.75120889863211;
const maxLat = 10.89935869248547;
const minLng = 122.2903781825292;
const maxLng = 122.4190035142424;

function lon2tile(lon, zoom) {
    return Math.floor((lon + 180) / 360 * Math.pow(2, zoom));
}
function lat2tile(lat, zoom) {
    return Math.floor((1 - Math.log(Math.tan(lat * Math.PI / 180) + 1 / Math.cos(lat * Math.PI / 180)) / Math.PI) / 2 * Math.pow(2, zoom));
}

const xMin = lon2tile(minLng, z);
const xMax = lon2tile(maxLng, z);
const yMin = lat2tile(maxLat, z); // maxLat has smaller y
const yMax = lat2tile(minLat, z);

let tilesToDownload = [];
for (let x = xMin; x <= xMax; x++) {
    for (let y = yMin; y <= yMax; y++) {
        tilesToDownload.push({ z, x, y });
    }
}

console.log(`Total tiles to download for zoom 17: ${tilesToDownload.length}`);

const baseDir = path.join(__dirname, 'static', 'tiles', z.toString());
if (!fs.existsSync(baseDir)) {
    fs.mkdirSync(baseDir, { recursive: true });
}

let activeDownloads = 0;
const MAX_CONCURRENT = 10;
let index = 0;

function downloadNext() {
    if (index >= tilesToDownload.length) {
        if (activeDownloads === 0) {
            console.log('All downloads finished.');
        }
        return;
    }

    const tile = tilesToDownload[index++];
    activeDownloads++;

    const xDir = path.join(baseDir, tile.x.toString());
    if (!fs.existsSync(xDir)) {
        fs.mkdirSync(xDir, { recursive: true });
    }

    const tilePath = path.join(xDir, `${tile.y}.png`);
    if (fs.existsSync(tilePath)) {
        activeDownloads--;
        downloadNext();
        return;
    }

    const url = `https://tile.openstreetmap.org/${tile.z}/${tile.x}/${tile.y}.png`;
    const options = {
        headers: { 'User-Agent': 'OfflineEvacuationApp/1.0' }
    };

    https.get(url, options, (res) => {
        if (res.statusCode === 200) {
            const fileStream = fs.createWriteStream(tilePath);
            res.pipe(fileStream);
            fileStream.on('finish', () => {
                fileStream.close();
                activeDownloads--;
                if (index % 100 === 0) console.log(`Downloaded ${index}/${tilesToDownload.length}`);
                downloadNext();
            });
        } else {
            console.error(`Failed to download ${url}: ${res.statusCode}`);
            activeDownloads--;
            downloadNext();
        }
    }).on('error', (err) => {
        console.error(`Error downloading ${url}: ${err.message}`);
        activeDownloads--;
        downloadNext();
    });
}

for (let i = 0; i < MAX_CONCURRENT; i++) {
    downloadNext();
}
